vmware-event-router/internal/processor/openfaas/openfaas.go

```diff
@@ -0,0 +1,322 @@
+package openfaas
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/avast/retry-go"
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	ofsdk "github.com/openfaas-incubator/connector-sdk/types"
+	"github.com/openfaas/faas-provider/auth"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/color"
+	config "github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/config/v1alpha1"
+	"github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/metrics"
+	"github.com/vmware-samples/vcenter-event-broker-appliance/vmware-event-router/internal/processor"
+)
+
+const (
+	defaultTopicDelimiter  = ","
+	defaultRebuildInterval = time.Second * 10
+	defaultTimeout         = time.Second * 15
+)
+
+var (
+	// ErrStopped error is returned when the processor has been shutdown but there
+	// are still inflight processing requests
+	ErrStopped = errors.New("processor already stopped")
+
+	// assert we implement Processor interface
+	_ processor.Processor = (*Processor)(nil)
+)
+
+type logger interface {
+	Printf(format string, v ...interface{})
+	Println(v ...interface{})
+}
+
+// invokeFunc is a function which invokes the given OpenFaaS function (fn) with
+// the specified message body. It returns the function response message, status
+// code, http headers and error.
+type invokeFunc func(ctx context.Context, fn string, message []byte) ([]byte, int, http.Header, error)
+
+// waitFunc is a wait function passed to waitForAll
+type waitFunc func(ctx context.Context) error
+
+// responseFunc implements ResponseSubscriber and is used to configure the
+// default response handler for the OpenFaaS processor
+type responseFunc func(ofsdk.InvokerResponse)
+
+// Response is a wrapper function to implement the OpenFaaS Response handler
+// interface
+func (r responseFunc) Response(res ofsdk.InvokerResponse) {
+	r(res)
+}
+
+// Processor implements the Processor interface
+type Processor struct {
+	controller ofsdk.Controller
+	ofsdk.ResponseSubscriber
+	respChan chan ofsdk.InvokerResponse // responses from sync fn invocation
+
+	// options
+	verbose         bool
+	topicDelimiter  string
+	rebuildInterval time.Duration
+	gatewayTimeout  time.Duration
+	// TODO (@embano1): make log interface for all processors/streams
+	*log.Logger
+
+	lock    sync.RWMutex
+	stats   metrics.EventStats
+	stopped bool
+}
+
+// NewProcessor returns an OpenFaaS processor for the given stream
+// source. Asynchronous function invocation can be configured for
+// high-throughput (non-blocking) requirements.
```

Comment:
These are very useful comments. Glad you have them and so many of the comments throughout the code.

Reply:
Thank you Frankie!